kafka-event-listener/src/main/java/org/hypertrace/core/kafka/event/listener/KafkaEventListener.java

```diff
@@ -0,0 +1,110 @@
+import java.util.concurrent.Executors;
+import java.util.function.BiConsumer;
+import lombok.Value;
+import org.apache.kafka.clients.consumer.Consumer;
+import org.apache.kafka.common.serialization.Deserializer;
+
+/**
+ * KafkaEventListener consumes events from a single Kafka Topic and on every message invokes
+ * provided callbacks. The callback invocation is done in a separate thread and needs to be
+ * concurrent safe.
+ */
+public class KafkaEventListener<K, V> implements AutoCloseable {
+  private final KafkaEventListenerThread<K, V> kafkaEventListenerThread;
+
+  private KafkaEventListener(KafkaEventListenerThread<K, V> kafkaEventListenerThread) {
+    this.kafkaEventListenerThread = kafkaEventListenerThread;
+  }
+
+  @Override
+  public void close() throws Exception {
+    kafkaEventListenerThread.interrupt();
+    kafkaEventListenerThread.join(Duration.ofSeconds(10).toMillis());
+  }
+
+  public static final class Builder<K, V> {
+    KafkaEventListenerThreadConfig<K, V> kafkaEventListenerThreadConfig;
+    List<BiConsumer<? super K, ? super V>> callbacks = new ArrayList<>();
+    ExecutorService executorService;
+
+    public Builder<K, V> addKafkaTopic(
+        String consumerName,
+        Config kafkaConfig,
+        Deserializer<K> keyDeserializer,
+        Deserializer<V> valueDeserializer) {
+      assertAbsenceOfKakfaTopic();
+      kafkaEventListenerThreadConfig =
+          new KafkaEventListenerThreadConfig<>(
+              consumerName, kafkaConfig, keyDeserializer, valueDeserializer, null, false);
+      return this;
+    }
+
+    public Builder<K, V> addKafkaConsumer(
+        String consumerName, Config kafkaConfig, Consumer<K, V> kafkaConsumer) {
+      assertAbsenceOfKakfaTopic();
+      kafkaEventListenerThreadConfig =
+          new KafkaEventListenerThreadConfig<>(
+              consumerName, kafkaConfig, null, null, kafkaConsumer, true);
+      return this;
+    }
+
+    public Builder<K, V> registerCallback(BiConsumer<? super K, ? super V> callbackFunction) {
+      callbacks.add(callbackFunction);
+      return this;
+    }
+
+    public Builder<K, V> withExecutorService(ExecutorService executorService) {
+      this.executorService = executorService;
+      return this;
+    }
+
+    public KafkaEventListener<K, V> build() {
+      if (Objects.isNull(kafkaEventListenerThreadConfig) || callbacks.isEmpty()) {
+        throw new IllegalArgumentException(
+            "one kafka topic and at least one callback are required for KafkaEventListener");
+      }
+      KafkaEventListenerThread<K, V> kafkaEventListenerThread =
+          kafkaEventListenerThreadConfig.getThreadFromConfig(callbacks);
+      if (Objects.isNull(executorService)) {
+        executorService = Executors.newSingleThreadExecutor();
+      }
+      executorService.submit(kafkaEventListenerThread);
+      return new KafkaEventListener<>(kafkaEventListenerThread);
+    }
+
+    private void assertAbsenceOfKakfaTopic() {
```

Comment:
Another way of making sure this is only done once is "promote" it to be an argument of the build method (similar to how the cacheloader is the arg to the build method in a guava cache).